src/coreclr/src/jit/lclvars.cpp

```diff
@@ -1858,7 +1851,26 @@ bool Compiler::StructPromotionHelper::CanPromoteStructVar(unsigned lclNum)
     }
 
     CORINFO_CLASS_HANDLE typeHnd = varDsc->lvVerTypeInfo.GetClassHandle();
-    return CanPromoteStructType(typeHnd);
+
+    bool canPromote = CanPromoteStructType(typeHnd);
+    if (canPromote && varDsc->lvIsMultiRegArgOrRet())
+    {
+        if (structPromotionInfo.fieldCnt > MAX_MULTIREG_COUNT)
+        {
+            canPromote = false;
+        }
+#ifdef UNIX_AMD64_ABI
+        else
+        {
+            SortStructFields();
+            if ((structPromotionInfo.fieldCnt == 2) && (structPromotionInfo.fields[1].fldOffset != TARGET_POINTER_SIZE))
+            {
+                canPromote = false;
```

Comment:
Yes, it will. I think it would be correct to always allow promotion if the number of fields is equal to the number of registers, but this seemed like a more conservative estimate to start from.